File: web/src/stores/authStore.js
// frontend/src/stores/authStore.js
import { create } from "zustand";
import { api } from "@/lib/api";

export const useAuthStore = create((set, get) => ({
    user: null,
    isLoading: true,
    isAuthenticated: false,
    isInitialized: false,

    initialize: () => {
        // Prevent double initialization
        if (get().isInitialized) return;

        const token = localStorage.getItem("access_token");
        const userJson = localStorage.getItem("user_data");

        if (token && userJson) {
            try {
                const user = JSON.parse(userJson);
                set({
                    user,
                    isAuthenticated: true,
                    isLoading: false,
                    isInitialized: true,
                });
            } catch {
                // Clear invalid data
                localStorage.removeItem("access_token");
                localStorage.removeItem("refresh_token");
                localStorage.removeItem("user_data");
                set({
                    user: null,
                    isAuthenticated: false,
                    isLoading: false,
                    isInitialized: true,
                });
            }
        } else {
            set({
                user: null,
                isAuthenticated: false,
                isLoading: false,
                isInitialized: true,
            });
        }
    },

    login: async (username, password) => {
        const data = await api.login({ username, password });
        localStorage.setItem("access_token", data.access_token);
        localStorage.setItem("refresh_token", data.refresh_token);
        const user = { username };
        localStorage.setItem("user_data", JSON.stringify(user));
        set({ user, isAuthenticated: true, isLoading: false });
        return data;
    },

    register: async (username, email, password) => {
        const data = await api.register({ username, email, password });
        localStorage.setItem("access_token", data.access_token);
        localStorage.setItem("refresh_token", data.refresh_token);
        const user = { username, email };
        localStorage.setItem("user_data", JSON.stringify(user));
        set({ user, isAuthenticated: true, isLoading: false });
        return data;
    },

    logout: () => {
        localStorage.removeItem("access_token");
        localStorage.removeItem("refresh_token");
        localStorage.removeItem("user_data");
        set({ user: null, isAuthenticated: false, isLoading: false });
    },
}));

File: mobile/lib/stores/auth_store.dart
// lib/stores/auth_store.dart
import 'dart:convert';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:justscroll/lib/api.dart';
import 'package:justscroll/models/user.dart';

class AuthState {
  final AppUser? user;
  final bool isLoading;
  final bool isAuthenticated;
  final bool isInitialized;

  const AuthState({
    this.user,
    this.isLoading = true,
    this.isAuthenticated = false,
    this.isInitialized = false,
  });

  AuthState copyWith({AppUser? user, bool? isLoading, bool? isAuthenticated, bool? isInitialized}) {
    return AuthState(
      user: user ?? this.user,
      isLoading: isLoading ?? this.isLoading,
      isAuthenticated: isAuthenticated ?? this.isAuthenticated,
      isInitialized: isInitialized ?? this.isInitialized,
    );
  }
}

class AuthNotifier extends StateNotifier<AuthState> {
  AuthNotifier() : super(const AuthState()) {
    initialize();
  }

  Future<void> initialize() async {
    if (state.isInitialized) return;
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString('access_token');
    final userJson = prefs.getString('user_data');

    if (token != null && userJson != null) {
      try {
        final user = AppUser.fromJson(jsonDecode(userJson));
        state = AuthState(user: user, isAuthenticated: true, isLoading: false, isInitialized: true);
      } catch (_) {
        await _clearStorage();
        state = const AuthState(isLoading: false, isInitialized: true);
      }
    } else {
      state = const AuthState(isLoading: false, isInitialized: true);
    }
  }

  Future<void> login(String username, String password) async {
    final data = await ApiClient.instance.login(username: username, password: password);
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('access_token', data['access_token']);
    await prefs.setString('refresh_token', data['refresh_token']);
    final user = AppUser(username: username);
    await prefs.setString('user_data', jsonEncode(user.toJson()));
    state = AuthState(user: user, isAuthenticated: true, isLoading: false, isInitialized: true);
  }

  Future<void> register(String username, String email, String password) async {
    final data = await ApiClient.instance.register(username: username, email: email, password: password);
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('access_token', data['access_token']);
    await prefs.setString('refresh_token', data['refresh_token']);
    final user = AppUser(username: username, email: email);
    await prefs.setString('user_data', jsonEncode(user.toJson()));
    state = AuthState(user: user, isAuthenticated: true, isLoading: false, isInitialized: true);
  }

  Future<void> logout() async {
    await _clearStorage();
    state = const AuthState(isLoading: false, isInitialized: true);
  }

  Future<void> _clearStorage() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove('access_token');
    await prefs.remove('refresh_token');
    await prefs.remove('user_data');
  }
}

final authStoreProvider = StateNotifierProvider<AuthNotifier, AuthState>((ref) => AuthNotifier());